### PR TITLE
perf!: Reduce Seedfinder heap allocation

### PR DESCRIPTION
**Pull request by paulgessinger**
The Seed finder does a large number of heap allocations for various different things. With this PR, I'm reducing the number of heap allocations considerably with little change to the public interface.

On my machine, the runtime of the `SeedingAlgorithm` on 10000 events with the generic detector drops from about 3s to 2s, for about a 33% speedup.

The number of heap allocations is reduced from 220 million to just over 300000 so about a reduction of >99%.

I checked that the estimated parameter ROOT files are identical by using our `compareRootFiles.C` script.

BREAKING CHANGE: The interface of a number of methods changes:
- `SeedFilter::filterSeeds_2SpFixed` (takes an output iterator, is now void)
- `SeedFilter::filterSeeds_1SpFixed` (takes an output iterator, is now void)
- `Seedfinder::createSeedsForGroup` (takes an output iterator, is now void, takes a state object as a reference)
- `Grid::collect` (returns `boost::container::small_vector<size_t, N>` where `N = 3^DIM` of the grid)

### CODE REVIEW

File: Core/include/Acts/Seeding/Seedfinder.ipp [hunk 1]
```diff
@@ -36,19 +36,17 @@ Seedfinder<external_spacepoint_t, platform_t>::Seedfinder(
 
 template <typename external_spacepoint_t, typename platform_t>
 template <typename sp_range_t>
-std::vector<Seed<external_spacepoint_t>>
-Seedfinder<external_spacepoint_t, platform_t>::createSeedsForGroup(
+void Seedfinder<external_spacepoint_t, platform_t>::createSeedsForGroup(
+    State& state,
+    std::back_insert_iterator<std::vector<Seed<external_spacepoint_t>>> outIt,
```
Comment (robertlangenberg) [inline Core/include/Acts/Seeding/Seedfinder.ipp:41]: Is it necessary to template this on a concrete type? Wouldn't it be more flexible (and more efficient?) to allow the user to provide a small_vector with around the expected number of SP preallocated?

Comment (paulgessinger) [reply Core/include/Acts/Seeding/Seedfinder.ipp:41]: I mean, the method is templated already, so in principle that should be possible I guess.

File: Examples/Framework/include/ActsExamples/EventData/ProtoTrack.hpp [hunk 2]
```diff
@@ -12,10 +12,12 @@
 
 #include <vector>
 
+#include <boost/container/small_vector.hpp>
+
 namespace ActsExamples {
 
 /// A proto track is a collection of hits identified by their indices.
-using ProtoTrack = std::vector<Index>;
+using ProtoTrack = boost::container::small_vector<Index, 10>;
```
Comment (robertlangenberg) [inline Examples/Framework/include/ActsExamples/EventData/ProtoTrack.hpp:20]: this uses 3x more space than necessary for 3 SP, right?

**Review by robertlangenberg — COMMENTED** [review]
The keyword for the changed interface needs to be in the PR description, right?

File: Core/include/Acts/Seeding/Seedfinder.ipp (hunk 1 above)
Comment (robertlangenberg) [reply Core/include/Acts/Seeding/Seedfinder.ipp:41]: Do you think that's overkill? You don't seem to be convinced :)

File: Examples/Framework/include/ActsExamples/EventData/ProtoTrack.hpp (hunk 2 above)
Comment (paulgessinger) [reply Examples/Framework/include/ActsExamples/EventData/ProtoTrack.hpp:20]: Yeah true, I just went with a low-ish number. Let's see what I get with 3.

Comment (paulgessinger) [reply Examples/Framework/include/ActsExamples/EventData/ProtoTrack.hpp]: In my testing, reducing to 3 appears to actually slow things down by about 3%, which I can't explain. I suggest we keep it at 10.

File: Core/include/Acts/Seeding/Seedfinder.ipp (hunk 1 above)
Comment (paulgessinger) [reply Core/include/Acts/Seeding/Seedfinder.ipp]: Done.

**Comment by paulgessinger** [comment]
> The keyword for the changed interface needs to be in the PR description, right?

Added.

**Comment by paulgessinger** [comment]
Clang seems to dislike the template template.

**Comment by paulgessinger** [comment]
Should compile now.

**Review by robertlangenberg — APPROVED** [review]

**Pull request merged**